2 as a player i want to be able to make a cocktail

NickoLefrench opened the pull request:
Player can make a cocktail, minigame makes score multiplier, minigame ends after timer and player can return to gameplay. Only way to manually quit however is to pause game entirely. Creating bug tracker to move onto next task

NickoLefrench committed:
Started Coktail Minigame

Fridges are different colours
Player can pick up and add to inventory colour specific ingredient

NickoLefrench committed:
Started Setup CocktailMinigame

Adjustments done to spacing of bar objects in scene
Started/setup of cocktail shaker minigame

NickoLefrench committed:
Progress on COcktailmingame

similar structure to beer
looking into rythm games

NickoLefrench committed:
Progress on Cocktail Minigame

Rythm game should be functional, Havent set the UI on the canvas yet
Starts to check if player has correct ingredient to give correct cocktail

NickoLefrench committed:
Finished Cocktail MiniGame!

Code should be completed, no written bugs seen
Canvas and UI still needs to be set up within the scene
Testing needed

NickoLefrench committed:
Settiung up game Objects for canvas

Creating empty game object and put them in the canvas group for the Cocktail Minigame UI

NickoLefrench committed:
Game starts but does not leave menu, null ref occurs

Cocktail minigame is breaking the UI management

NickoLefrench committed:
fixed stash conflict

NickoLefrench committed:
Cocktailminigame IS functional

game launches, part of the cocktail UI is visible in gameplay

NickoLefrench committed:
Finished Cocktail Minigame code

Player should be able to enter and exit cocktail minigame
code has prooper logic
updates to GameManager

NickoLefrench committed:
CkctailMinigame officially funtional and was tested

player can interact with minigame and it starts and exits properly
Cocktail and arrows are misplace in scene :/

NickoLefrench committed:
Fixing Cocktail Shaker in the scene + code so its playable

Player can interact with cocktail shaker starting minigame
Only right pressed and unpressed arrows flash
Score multiplier works

NickoLefrench committed:
Adjusting to alternating arrrows

Only left or right flashes during minigame
made alternation more robust
still need to adjust scene hierachy?

NickoLefrench committed:
Player can play the cocktail minigame

Cocktail minigame complete
Player plays it until timer runs out
both arrows flash
can only quit by pausing game

vanyamil commented on Assets/Scripts/Interactables/CocktailFridge.cs:
```diff
@@ -1,13 +1,102 @@
+        private ColorCategory GetColorCategory(Color color) //maps the Color to a ColorCategory enum predifined
+        {
+            // Here we compare colors with a small tolerance for floating-point precision issues
+            if (ColorsAreClose(color, new Color(0.934f, 0f, 0f, 1f)))
```
Set these as constants, either in code (final or static), or through some kind of prefab.

NickoLefrench replied:
The Colour category as constants? From what i understand from the previous comment, if i rework the class so that its the colour category of the fridge then i would set it as a constant

NickoLefrench replied:
I believe i resolved it and make them constants in the rework i just did setting colourtype as a category

vanyamil commented on Assets/Scripts/Interactables/CocktailFridge.cs:
```diff
@@ -1,13 +1,102 @@
+            return ColorCategory.Unknown;
+        }
+
+        private bool ColorsAreClose(Color color1, Color color2, float tolerance = 0.1f) //some math from chatgpt to create a window of error of the adjusted colors on the prefab
```
Comments on the side but aligned with spaces hurt my brain :D just put them in the body of the code, on the lines before return

vanyamil replied:
Also, why do we use the Color of the fridge instead of say setting a ColorCategory on the Fridge?

NickoLefrench replied:
I set the colour of the fridge before hand and formulated the class to check/determine/compare what colour the current fridge is and given the corresponding item

NickoLefrench replied:
Iw ill adjust all the comments too

NickoLefrench replied:
updated comments to be o nthe lines before corresponding code, except for defining variables, i adjusted so its cleaner still

NickoLefrench replied:
I defined to get the corresponding ingredient with  the Color of the fridge as a type within the enum

vanyamil commented on Assets/Scripts/Interactables/CocktailFridge.cs:
```diff
@@ -1,13 +1,102 @@
+
+            if (FridgeSpriteRenderer != null)
+            {
+                Color fridgeColor = FridgeSpriteRenderer.color;
```
Repeating comment from elsewhere: 
why do we use the Color of the fridge instead of say setting a ColorCategory on the Fridge?

vanyamil replied:
In fact, if you design it that way (Fridge has an enum value for what it can return, or an item definition), it will be more extendable in the future.

NickoLefrench replied:
Yeeeaaaah  i should definetly do that so the logic wont be so fragile

NickoLefrench replied:
As previously said, but i think i misexplained, i have just set the colorCategory of the fridge to get the corresponding item type in an enum so its more flexible and less fragile

vanyamil commented on Assets/Scripts/Interactables/CocktailMiniGame.cs:
```diff
@@ -0,0 +1,374 @@
+    {
+
+        // References to the sprites for the arrows
+        public Sprite leftUIPressed;
```
It's still our first game, but read up on public vs [SerializeField] private.

NickoLefrench replied:
Oh right, if i serialize private, i can still adjust it in the inspector and halt other scripts from accessing it. Ill start ensuring that all my variables are private unless another script really needs access it and serialize variable when need to adjust in the inspector

NickoLefrench replied:
Done, almost all of the public variables are now private serialized

vanyamil commented on Assets/Scripts/Interactables/CocktailMiniGame.cs:
```diff
@@ -0,0 +1,374 @@
+        public GameObject leftArrow; // Assign the LeftArrow object here
+        public GameObject rightArrow; // Assign the RightArrow object here
+
+        private PlayerInteraction player; // Store player reference
```
Try to separate into sections things that are public/serializable (aka modifiable in editor) from private (aka only modifiable within this class's code)

NickoLefrench replied:
I can definetly, and most definetly should, privatizze and serialize all of the variables used

NickoLefrench replied:
I cleaned up the variables and comments to be less brain hurting

vanyamil commented on Assets/Scripts/Interactables/CocktailMiniGame.cs:
```diff
@@ -0,0 +1,374 @@
+        private bool detectedInputDuringMiniGame = false;
+        public Item itemToSpawn;
+
+        public KeyCode leftKey = KeyCode.A;  // Key for left arrow
```
I strongly recommend switching to Axis instead.

NickoLefrench replied:
I could rework it use GetAxis Instead, as from what i understand its more versatile with controllers and adjustable from the input manager of the project settings. Which we will need to install the Input System Package according to Unity Docs as 'For new projects you should use the newer and Input System Package'?

NickoLefrench replied:
switched to GetAxis, had to rework almost everything.
- The arrow switching and sprite handling logic now directly depend on a centralized mechanism so that it would integrate better with GetAxis
- - The SwitchArrowsCoroutine now alternates a leftArrowActive boolean on a timed interval,. I tired to make sure its regular and predictable switching between the left and right arrows
- _Get Axis should (keyword) be syncronized with the arrow sprites changing to affect the score multipler

it has not been tested tho

vanyamil commented on Assets/Scripts/Interactables/CocktailMiniGame.cs:
```diff
@@ -0,0 +1,374 @@
+
+        public CanvasGroup cocktailCanvasGroup; // Reference to your Canvas Group for the mini-game UI
+
+        [Tooltip("Duration of the mini-game in seconds.")]
```
Tooltips are great, good for self-documentation!

vanyamil commented on Assets/Scripts/Interactables/CocktailMiniGame.cs:
```diff
@@ -0,0 +1,374 @@
+            // Set the Cocktail Shaker to be visible at the start
+            if (cocktailShakerRenderer != null)
+            {
+                cocktailShakerRenderer.enabled = true;
```
Should this instead be .SetEnabled() ?

vanyamil replied:
or SetActive?

NickoLefrench replied:
cocktailShakerRenderer.enabled enables the visibility of just the sprite as .enabled toggles specific compenents. SetActive would toggle the entire object and all of its components

vanyamil commented on Assets/Scripts/Interactables/CocktailMiniGame.cs:
```diff
@@ -0,0 +1,374 @@
+        // Update is called once per frame
+        void Update()
+        {
+           /* if (isMiniGameActive && Input.GetKeyDown(Quit))       LEGACY
```
No need to keep code commented out - we have version control

NickoLefrench replied:
just had it for legacy sake as i was struggling to have the player manually quit and tried different things

NickoLefrench replied:
I did leave a bit of legacy code that needs to be cleaned up since i rehashed big chunks, but i will delete it tmr with i adjust tunable handler

vanyamil commented on Assets/Scripts/Interactables/CocktailMiniGame.cs:
```diff
@@ -0,0 +1,374 @@
+            }
+
+            // Ensure arrows are hidden until the minigame starts
+            if (leftArrow != null) leftArrow.SetActive(false);
```
If you do the non-null check at Awake instead of at Start (I think?), you could also do a Debug.LogWarning saying "hey dont forget  to assign this variable"

NickoLefrench replied:
I could make an awake class that double checks if anything is null

NickoLefrench replied:
awake null check added

vanyamil commented on Assets/Scripts/Interactables/CocktailMiniGame.cs:
```diff
@@ -0,0 +1,374 @@
+            detectedInputDuringMiniGame = false;
+
+            // Hide or show relevant elements for the minigame
+            if (cocktailShakerRenderer != null)
```
Is it even a mini game if the renderer isn't there?

NickoLefrench replied:
What do you mean? the sprite renderer?

vanyamil commented on Assets/Scripts/Interactables/CocktailMiniGame.cs:
```diff
@@ -0,0 +1,374 @@
+            rhythmCoroutine = StartCoroutine(SwitchArrowsCoroutine());
+
+            // Start the timer coroutine for automatic end
+            StartCoroutine(MiniGameTimer());
```
Is starting multiple coroutines on same game object legal in Unity?

NickoLefrench replied:
there is no legal limit of coroutines that you can have. It can be tricky from an architectual stand point to have many nested coroutines running as it would be harder to debug, but its prossible

Check out this guy testing the myth by having a million coroutines running, its possible, but the more present, the higher possibility of framerate issues and debugging complexity

https://gamedev.stackexchange.com/questions/205124/are-there-limits-to-the-number-of-coroutines-in-unity

vanyamil commented on Assets/Scripts/Interactables/CocktailMiniGame.cs:
```diff
@@ -0,0 +1,374 @@
+
+                // Switch the sprite based on active arrow
+                //Sets the sprite property of the Image component on each arrow to either the pressed or unpressed state
+                if (leftArrowActive)
```
null checks should no longer be needed. We should confirm null checks early, and if anything still null, just not let minigame happen.

NickoLefrench replied:
awake null check should be good right?

vanyamil commented on Assets/Scripts/Interactables/CocktailMiniGame.cs:
```diff
@@ -0,0 +1,374 @@
+        {
+
+            // Increase multiplier for a perfect hit
+            if (timeSinceSwitch <= perfectHitWindow)
```
This implies I can spam left-right and keep getting perfect scores? should there be a lower bound?

NickoLefrench replied:
from what ive tested, its fairly precise, you can only see your score multiplier go up when you time your button press on UIPressed.  HOwever we may need to tweek the timeframe and precision as well as a more noticeable cue that you timed it correctly

vanyamil commented on Assets/Scripts/Interactables/CocktailMiniGame.cs:
```diff
@@ -0,0 +1,374 @@
+            // Increase multiplier for a perfect hit
+            if (timeSinceSwitch <= perfectHitWindow)
+            {
+                scoreMultiplier += 0.5f; 
```
These values should be tunable, rather than set in stone within a function. At bare minimum, move these into serialized values adjustable in editor, or we could look at the TunableHandler/similar ideas.

NickoLefrench replied:
fair enough, i wasnt sure to touch tunable, but your probably right it could be hadled there, especially if we want it to affect the score from picking up the patrons coins

NickoLefrench replied:
I have yet to connect score multiplier with tunable handler, that should be a quick fix tmr

NickoLefrench replied:
i adjusted it to be a part of tunable handler, but TunableHandler.Instance is inaccessible due to protection level, should i be adjusting tunable get/setter, or define a different accessor from cocktailminigame?

vanyamil commented on Assets/Scripts/Interactables/CocktailMiniGame.cs:
```diff
@@ -0,0 +1,374 @@
+                //remove  Red and Blue ingredients from the players inventory
+                player.CurrentInventory.RemoveFirstItemOfType(Item.ItemType.YellowIngredient);
+                player.CurrentInventory.RemoveFirstItemOfType(Item.ItemType.BlueIngredient);
+            }
```
else?

NickoLefrench replied:
yeeeaaaaahh i should make this a switch

NickoLefrench replied:
forgto to make this a switch, will work on

vanyamil commented on Assets/Scripts/Interactables/CocktailMiniGame.cs:
```diff
@@ -0,0 +1,374 @@
+            GameManager.Instance.EndCocktailMiniGame();
+        }
+
+        private IEnumerator MiniGameTimer()
```
I kinda super dig this solution!

NickoLefrench replied:
thank you :)

vanyamil reviewed (changes requested):
Decent progress, but still things to do. Don't forget the merge too!

NickoLefrench committed:
Updated comments and get ingredient by colour type

Updated comments to be clean and on the lines before corresponding code
Adjusted fridges to get the corresponding ingredients in regarde to the fridges colour type

NickoLefrench committed:
Big cockyail minigame update

Reworked input to implement GetAxis
Serialized private all the public variables
awake function null check
arrow swithcing managed by SelUIVisibility

NickoLefrench committed:
adjusted scoremultiplier

scoremultiplier part of tunablehandler, not working yet

NickoLefrench committed:
Nothing really changed

trying different solutions for attaching scoremultiplier to tunable handler
was distracted so i didnt get far

NickoLefrench committed:
thinkerig in tunable

slow proge=ress

vanyamil committed:
Revert "thinkerig in tunable"

This reverts commit c99fa41a3205c16b8d7ed04f390934a34883d195.

vanyamil committed:
Revert "Nothing really changed"

This reverts commit 17741b003035b72b837c5a8d495883b51b7a9f0f.

vanyamil committed:
Revert "adjusted scoremultiplier"

This reverts commit 06e51bcb8da403b047a652e0c588cf31c9a75bb8.

NickoLefrench committed:
cleaned up legacy comments

removed unecessary comments

NickoLefrench committed:
new working update

Scoremultiplier affects score directly
GetAxisRaw

NickoLefrench committed:
Cocktail Minigame somewaht functional

Scormultipler affects score
Player can pick ingredients and give corresponding cocktail
Wohhoo

NickoLefrench committed:
Merge branch 'main' into 2-as-a-player-i-want-to-be-able-to-make-a-cocktail

NickoLefrench committed:
Merge Resolve

Merged main into cocktailminigame

Had to adjust GameManager to work with the rest of the game

Scene needs to be re adjusted

NickoLefrench committed:
Scene adjustments

Adjusted scene and fridges
cant remember how i set up the UI

NickoLefrench committed:
Cocktail minigame fully finished and functional in game

cocktail minigame works
patrons can order cocktail or beer
scormultiplier affects score received after picking up coins

vanyamil committed:
Final cleanup
